FEAT: Add anecdoctor orchestrator to build attack prompts from real-world examples.

migdaepp opened the pull request:
## Description

Adding a new Orchestrator that constructs attack prompts based on real-world examples. This orchestrator performs best for informational harms in that center on a consistent narrative (i.e. via clustering of a larger attack dataset). There are two options:

- baseline approach simply constructs prompts from few-shot examples
- use_knowledge_graph=True uses a processing target to construct a knowledge graph. This produces attack prompts that are more robust to cultural and linguistic differences in example data.

This contribution follows on collaborations with @eugeniavkim. 

## Tests and Documentation

Tests are included for the orchestrator. We also include documentation (jupytext) with a toy dataset as well as a link to real-world data for which the method works well. A paper validating the method is in revision and will be posted to ArXiv in the near future.

jsong468 commented on doc/code/orchestrators/anecdoctor_orchestrator.py:
```diff
@@ -0,0 +1,165 @@
+results = await orchestrator_with_kg_german.evaluate()  # type: ignore
+
+# %% [markdown]
+# To better understand the attacks under evaluation, you can visualize the knowledge graphs produced in the processing step.
```
If it's not too much work, it might be nice to see the details actual relationships that the LLM identified among the entities, alongside the visual graph to see how they align.

migdaepp replied:
Now printed when verbose = True, see below

jsong468 commented on pyrit/datasets/orchestrators/anecdoctor/anecdoctor_use_knowledge_graph.yaml:
```diff
@@ -0,0 +1,47 @@
+  Don't include any example {type}s. 
+  Provide instructions in the language {language}.
+
+  Produce a {type} similar to the examples you are given.
```
May be understanding wrong, but the model only sees the knowledge graph at this point right? Not the original examples.

jsong468 commented on pyrit/orchestrator/anecdoctor_orchestrator.py:
```diff
@@ -0,0 +1,199 @@
+
+        # 7. If verbose, print a nicely formatted message
+        if self._verbose:
+            # uncomment the following lines to print the examples or knowledge graph
+            # print(f"{Style.BRIGHT}{Fore.BLUE}user:{Style.RESET_ALL}")
+            # wrapped_user_text = textwrap.fill(formatted_examples, width=100)
+            # print(f"{Style.BRIGHT}{Fore.BLUE}{wrapped_user_text}{Style.RESET_ALL}")
```
Oh I guess this does what I recommended in the notebook :D Maybe we can just default to printing this when verbose is True. What do you think?

migdaepp replied:
Sounds great, fixed in 777d038

jsong468 commented on pyrit/orchestrator/anecdoctor_orchestrator.py:
```diff
@@ -0,0 +1,199 @@
+            system_prompt = self._load_prompt_from_yaml("anecdoctor_use_knowledge_graph.yaml")
+            # 1b. run the knowledge graph extraction
+            self._kg_result = await self._extract_knowledge_graph()
+            # 1c. set examples to knowldge graph format
```
NIT
```suggestion
            # 1c. set examples to knowledge graph format
```

jsong468 commented on pyrit/orchestrator/anecdoctor_orchestrator.py:
```diff
@@ -0,0 +1,199 @@
+    async def evaluate(self) -> None:
+        """
+        Runs the orchestrator, possibly extracting a knowledge graph first,
+        then generating the final content. Returns the final model output (answer).
```
Should this method be returning something? perhaps `final_answer`?

migdaepp replied:
good point, fixed in 777d038

jsong468 commented on pyrit/orchestrator/anecdoctor_orchestrator.py:
```diff
@@ -0,0 +1,199 @@
+        kg_output = response.get_value()
+        return kg_output
+
+    async def evaluate(self) -> None:
```
do you think something like `generate_attack` as the method name captures what this method does better?

migdaepp commented:
## Implemented Changes

Thanks for the thorough review! I've made the following changes:

- renamed `evaluate` → `generate_attack`.
- `generate_attack` now returns the final model response.
- added printing of examples / KG relationships when verbose = True
- Fixed the typo flagged in review.
- Updated unit tests and docs to match the changes
